rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java

```diff
@@ -460,46 +575,149 @@ private boolean isJaxRsCancel(ContainerRequestContext requestContext, ContainerR
         return false;
     }
 
-    private URI startLRA(URI parentLRA, Method method, Long timeout) {
-        // timeout should already have been converted to milliseconds
-        String clientId = method.getDeclaringClass().getName() + "#" + method.getName();
+    // the request filter may perform multiple and in failure scenarios the LRA may be left in an ambiguous state:
+    // the following structure is used to track progress so that such failures can be reported in the response
+    // filter processing
+    private enum ProgressStep {
+        Left ("leave succeeded"),
+        LeaveFailed("leave failed"),
+        Started("start ok"),
+        StartFailed("start failed"),
+        Joined("join succeeded"),
+        JoinFailed("join failed"),
+        Ended("end ok"),
+        CloseFailed("close failed"),
+        CancelFailed("cancel failed");
+
+        final String status;
+
+        ProgressStep(final String status) {
+            this.status = status;
+        }
 
-        return lraClient.startLRA(parentLRA, clientId, timeout, ChronoUnit.MILLIS);
+        @Override
+        public String toString() {
+            return status;
+        }
     }
 
-    private void resumeTransaction(URI lraId) {
-        // nothing to do
+    // list of steps (both successful and unsuccesful) performed so far by the request and response filter
+    // and is used for error reporting
+    private static class Progress {
+        static EnumSet<ProgressStep> failures = EnumSet.of(
```

Comment:
Would no this `EnumSet` be more appropriate to be part (a member) of the `ProgressStep` Enum itself?

Reply:
Not necessarily. Progress knows what constitutes a failure. Again, if we want this code to be perfect then different engineers have different concepts of perfection and such a strategy just serves to prolong the process.